lib/auth/auth.js

```diff
@@ -26,117 +30,150 @@ auth.setAuthHandler = handler => {
+        method = 'query';
+        version = 'v2';
     } else if (data['X-Amz-Algorithm']) {
-        log.trace('authenticating request with Auth v4 using query string');
-        return authV4.queryAuthCheck.check(request, log, data);
+        method = 'query';
+        version = 'v4';
     }
-    // If no auth information is provided in request, then
-    // user is part of 'All Users Group' so send back this
-    // group as the canonicalID
-    log.debug('No authentication provided. User identified as public');
-    const authInfo = new AuthInfo({ canonicalID: constants.publicId });
-    return { err: null, data: authInfo };
-};
 
-auth.doAuth = (request, log, cb, awsService, requestContext) => {
-    const res = auth.check(request, log, awsService, request.query);
-    requestContext.setAuthType(res.authType);
-    requestContext.setSignatureVersion(res.signatureVersion);
-    requestContext.setSignatureAge(res.signatureAge);
+    // Here, either both values are set, or none is set
+    if (version !== null && method !== null) {
+        if (!checkFunctions[version] || !checkFunctions[version][method]) {
+            log.trace('invalid auth version or method',
+                      { version, authMethod: method });
+            return { err: errors.NotImplemented };
+        }
+        log.trace('identified auth method', { version, authMethod: method });
+        return checkFunctions[version][method](request, log, data, awsService);
+    }
+
+    // no auth info identified
+    log.debug('assuming public user');
+    return { err: null, params: publicUserInfo };
+}
+
+function doAuth(request, log, cb, awsService, requestContext) {
+    const res = extractParams(request, log, awsService, request.query);
     if (res.err) {
         return cb(res.err);
-    } else if (res.version === 2) {
-        return vault.authenticateV2Request(res.data.accessKey,
-            res.data.signatureFromRequest,
-            res.data.stringToSign, res.data.algo, requestContext, log,
-            (err, authInfo) => {
-                if (err) {
-                    return cb(err);
-                }
-                return cb(null, authInfo);
-            });
-    } else if (res.version === 4) {
-        res.data.log = log;
-        return vault.authenticateV4Request(res.data, requestContext,
-            cb, awsService);
-    } else if (res.data instanceof AuthInfo) {
-        return cb(null, res.data);
+    } else if (res.params instanceof AuthInfo) {
+        return cb(null, res.params);
     }
+
+    requestContext.setAuthType(res.params.data.authType);
+    requestContext.setSignatureVersion(res.params.data.signatureVersion);
+    requestContext.setSignatureAge(res.params.data.signatureAge);
+
+    // Corner cases managed, we're left with normal auth
+    res.params.log = log;
+    if (res.params.version === 2) {
+        return vault.authenticateV2Request(res.params, requestContext, cb);
+    } else if (res.params.version === 4) {
```

Comment:
yes but I have to keep the if for having the possibility to return and internal error
```js
if(res.params.version !== 4 && res.params.version !== 2)
```